File: push/src/main/java/com/xiaomi/push/service/XMPushServiceChannelJobs.java
package com.xiaomi.push.service;

import android.text.TextUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.stats.StatsHelper;
import com.xiaomi.smack.Connection;
import com.xiaomi.smack.XMPPException;

final class BindJob extends XMPushService.Job {
    private final XMPushService service;
    PushClientsManager.ClientLoginInfo loginInfo;

    BindJob(XMPushService xMPushService, PushClientsManager.ClientLoginInfo clientLoginInfo) {
        super(9);
        this.service = xMPushService;
        this.loginInfo = clientLoginInfo;
    }

    @Override
    public String getDesc() {
        return "bind the client. " + this.loginInfo.chid;
    }

    @Override
    public void process() {
        try {
            if (this.service.isConnected()) {
                PushClientsManager.ClientLoginInfo clientLoginInfoByChidAndUserId = PushClientsManager.getInstance().getClientLoginInfoByChidAndUserId(this.loginInfo.chid, this.loginInfo.userId);
                if (clientLoginInfoByChidAndUserId == null) {
                    MyLog.w("ignore bind because the channel " + this.loginInfo.chid + " is removed ");
                } else if (clientLoginInfoByChidAndUserId.status == PushClientsManager.ClientStatus.unbind) {
                    clientLoginInfoByChidAndUserId.setStatus(PushClientsManager.ClientStatus.binding, 0, 0, null, null);
                    Connection currentConnection = this.service.getCurrentConnection();
                    if (currentConnection != null) {
                        currentConnection.bind(clientLoginInfoByChidAndUserId);
                        StatsHelper.statsBind(this.service, clientLoginInfoByChidAndUserId);
                    }
                } else {
                    MyLog.w("trying duplicate bind, ingore! " + clientLoginInfoByChidAndUserId.status);
                }
            } else {
                MyLog.e("trying bind while the connection is not created, quit!");
            }
        } catch (Exception e) {
            MyLog.e(e);
            this.service.disconnect(10, e);
        } catch (Throwable unused) {
        }
    }
}

final class BindTimeoutJob extends XMPushService.Job {
    public static final int BIND_TIMEOUT = 60000;
    private final PushClientsManager.ClientLoginInfo loginInfo;

    BindTimeoutJob(PushClientsManager.ClientLoginInfo clientLoginInfo) {
        super(12);
        this.loginInfo = clientLoginInfo;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof BindTimeoutJob) {
            return TextUtils.equals(((BindTimeoutJob) obj).loginInfo.chid, this.loginInfo.chid);
        }
        return false;
    }

    @Override
    public String getDesc() {
        return "bind time out. chid=" + this.loginInfo.chid;
    }

    @Override
    public int hashCode() {
        return this.loginInfo.chid.hashCode();
    }

    @Override
    public void process() {
        this.loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, 1, 21, null, null);
    }
}

final class PingJob extends XMPushService.Job {
    private final XMPushService service;
    boolean isPong;

    PingJob(XMPushService xMPushService) {
        super(4);
        this.service = xMPushService;
    }

    PingJob(XMPushService xMPushService, boolean z) {
        super(4);
        this.service = xMPushService;
        this.isPong = z;
    }

    @Override
    public String getDesc() {
        return "send ping..";
    }

    @Override
    public void process() {
        if (this.service.isConnected()) {
            try {
                if (!this.isPong) {
                    StatsHelper.pingStarted();
                }
                Connection currentConnection = this.service.getCurrentConnection();
                if (currentConnection != null) {
                    currentConnection.ping(this.isPong);
                }
            } catch (XMPPException e) {
                MyLog.e(e);
                this.service.disconnect(10, e);
            }
        }
    }
}

final class ReBindJob extends XMPushService.Job {
    private final XMPushService service;
    PushClientsManager.ClientLoginInfo loginInfo;

    ReBindJob(XMPushService xMPushService, PushClientsManager.ClientLoginInfo clientLoginInfo) {
        super(4);
        this.service = xMPushService;
        this.loginInfo = clientLoginInfo;
    }

    @Override
    public String getDesc() {
        return "rebind the client. " + this.loginInfo.chid;
    }

    @Override
    public void process() {
        try {
            this.loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, 1, 16, null, null);
            Connection currentConnection = this.service.getCurrentConnection();
            if (currentConnection != null) {
                currentConnection.unbind(this.loginInfo.chid, this.loginInfo.userId);
                this.loginInfo.setStatus(PushClientsManager.ClientStatus.binding, 1, 16, null, null);
                currentConnection.bind(this.loginInfo);
            }
        } catch (XMPPException e) {
            MyLog.e(e);
            this.service.disconnect(10, e);
        }
    }
}

final class UnbindJob extends XMPushService.Job {
    private final XMPushService service;
    String kickType;
    PushClientsManager.ClientLoginInfo loginInfo;
    int notifyType;
    String reason;

    UnbindJob(XMPushService xMPushService, PushClientsManager.ClientLoginInfo clientLoginInfo, int i, String str, String str2) {
        super(9);
        this.service = xMPushService;
        this.loginInfo = clientLoginInfo;
        this.notifyType = i;
        this.kickType = str;
        this.reason = str2;
    }

    @Override
    public String getDesc() {
        return "unbind the channel. " + this.loginInfo.chid;
    }

    @Override
    public void process() {
        if (this.loginInfo.status != PushClientsManager.ClientStatus.unbind && this.service.getCurrentConnection() != null) {
            try {
                this.service.getCurrentConnection().unbind(this.loginInfo.chid, this.loginInfo.userId);
            } catch (XMPPException e) {
                MyLog.e(e);
                this.service.disconnect(10, e);
            }
        }
        this.loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, this.notifyType, 0, this.reason, this.kickType);
    }
}

File: push/src/main/java/com/xiaomi/push/service/XMPushServiceChannelJobs.kt
package com.xiaomi.push.service

import android.text.TextUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.stats.StatsHelper
import com.xiaomi.smack.Connection
import com.xiaomi.smack.XMPPException

class BindJob(
    private val service: XMPushService,
    val loginInfo: PushClientsManager.ClientLoginInfo,
) : XMPushService.Job(XMPushServiceJob.TYPE_BIND_UNBIND) {
    override fun getDesc(): String = "bind the client. ${loginInfo.chid}"

    override fun process() {
        try {
            if (!service.isConnected) {
                MyLog.e("trying bind while the connection is not created, quit!")
                return
            }
            val activeClient = PushClientsManager.getInstance().getClientLoginInfoByChidAndUserId(loginInfo.chid, loginInfo.userId)
            when {
                activeClient == null -> MyLog.w("ignore bind because the channel ${loginInfo.chid} is removed ")
                activeClient.status == PushClientsManager.ClientStatus.unbind -> {
                    activeClient.setStatus(PushClientsManager.ClientStatus.binding, 0, 0, null, null)
                    service.currentConnection?.let { connection ->
                        connection.bind(activeClient)
                        StatsHelper.statsBind(service, activeClient)
                    }
                }
                else -> MyLog.w("trying duplicate bind, ingore! ${activeClient.status}")
            }
        } catch (e: Exception) {
            MyLog.e(e)
            service.disconnect(10, e)
        } catch (_: Throwable) {
        }
    }
}

class BindTimeoutJob(
    private val loginInfo: PushClientsManager.ClientLoginInfo,
) : XMPushService.Job(XMPushServiceJob.TYPE_BIND_TIMEOUT) {
    companion object {
        const val BIND_TIMEOUT = 60000
    }

    override fun equals(other: Any?): Boolean {
        return other is BindTimeoutJob && TextUtils.equals(other.loginInfo.chid, loginInfo.chid)
    }

    override fun hashCode(): Int = loginInfo.chid.hashCode()

    override fun getDesc(): String = "bind time out. chid=${loginInfo.chid}"

    override fun process() {
        loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, 1, PushConstants.ERROR_BIND_TIMEOUT, null, null)
    }
}

class PingJob(
    private val service: XMPushService,
    private val isPong: Boolean = false,
) : XMPushService.Job(XMPushServiceJob.TYPE_SEND_MSG) {
    override fun getDesc(): String = "send ping.."

    override fun process() {
        if (!service.isConnected) {
            return
        }
        try {
            if (!isPong) {
                StatsHelper.pingStarted()
            }
            service.currentConnection?.ping(isPong)
        } catch (e: XMPPException) {
            MyLog.e(e)
            service.disconnect(10, e)
        }
    }
}

class ReBindJob(
    private val service: XMPushService,
    val loginInfo: PushClientsManager.ClientLoginInfo,
) : XMPushService.Job(XMPushServiceJob.TYPE_SEND_MSG) {
    override fun getDesc(): String = "rebind the client. ${loginInfo.chid}"

    override fun process() {
        try {
            loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, 1, 16, null, null)
            service.currentConnection?.let { connection ->
                connection.unbind(loginInfo.chid, loginInfo.userId)
                loginInfo.setStatus(PushClientsManager.ClientStatus.binding, 1, 16, null, null)
                connection.bind(loginInfo)
            }
        } catch (e: XMPPException) {
            MyLog.e(e)
            service.disconnect(10, e)
        }
    }
}

class UnbindJob(
    private val service: XMPushService,
    val loginInfo: PushClientsManager.ClientLoginInfo,
    private val notifyType: Int,
    private val kickType: String?,
    private val reason: String?,
) : XMPushService.Job(XMPushServiceJob.TYPE_BIND_UNBIND) {
    override fun getDesc(): String = "unbind the channel. ${loginInfo.chid}"

    override fun process() {
        if (loginInfo.status != PushClientsManager.ClientStatus.unbind && service.currentConnection != null) {
            try {
                service.currentConnection!!.unbind(loginInfo.chid, loginInfo.userId)
            } catch (e: XMPPException) {
                MyLog.e(e)
                service.disconnect(10, e)
            }
        }
        loginInfo.setStatus(PushClientsManager.ClientStatus.unbind, notifyType, 0, reason, kickType)
    }
}
